mediagrains/grains/BaseGrain.py

```diff
@@ -0,0 +1,630 @@
+from enum import IntEnum
+from uuid import UUID
+from mediatimestamp.immutable import (
+    Timestamp,
+    TimeOffset,
+    TimeRange,
+    SupportsMediaTimeOffset,
+    SupportsMediaTimestamp,
+    mediatimeoffset,
+    mediatimestamp)
+from collections.abc import Sequence, MutableSequence, Mapping
+from fractions import Fraction
+from copy import copy, deepcopy
+from inspect import isawaitable
+
+from typing import (
+    List,
+    Dict,
+    Any,
+    TypeAlias,
+    Union,
+    SupportsBytes,
+    Optional,
+    overload,
+    Tuple,
+    cast,
+    Sized,
+    Iterator,
+    Iterable,
+    Awaitable,
+    Generator)
+
+from ..typing import (
+    RationalTypes,
+    GrainMetadataDict,
+    GrainDataType,
+    EmptyGrainMetadataDict,
+    FractionDict,
+    TimeLabel,
+    GrainDataParameterType)
+
+
+class GrainType(IntEnum):
+    GRAIN = 0,
+    VIDEOGRAIN = 1,
+    AUDIOGRAIN = 2,
+    CODEDVIDEOGRAIN = 3,
+    CODEDAUDIOGRAIN = 4,
+    EVENTGRAIN = 5
+
+
+def _stringify_timestamp_input(value: Union[SupportsMediaTimestamp, SupportsMediaTimeOffset, str]) -> str:
+    if isinstance(value, SupportsMediaTimestamp):
+        value = mediatimestamp(value).to_sec_nsec()
+    elif isinstance(value, SupportsMediaTimeOffset):
+        value = mediatimeoffset(value).to_sec_nsec()
+    elif not isinstance(value, str):
+        raise ValueError(
+            f"{repr(value)} is not a type that can be converted to a Timestamp or TimeOffset, nor is it a string.")
+
+    return value
+
+
+# For use in 'new style' grain init. Can be removed once old style is excised.
+def new_attributes_for_grain_type(grain_type: str | GrainType):
+    return(attributes_for_grain_type(grain_type=grain_type, new=True))
+
+
+# As above, the 'new' parameter will give 'new style' grain init info. 'x if not new else' statements can be removed
+# once 'old style' is excised.
+def attributes_for_grain_type(grain_type: str | GrainType, new: bool = False) -> List[str]:
+    """Returns a list of attributes for a particular grain type. Useful for testing."""
+    # cast to the correct string
+    if type(grain_type) is GrainType:
+        grain_type = ["grain", "video", "audio", "coded_video", "coded_audio", "event"][grain_type]
+
+    COMMON_ATTRS = ['source_id' if not new else "src_id", 'flow_id', 'origin_timestamp', 'sync_timestamp',
+                    'creation_timestamp', 'rate', 'duration']
+    # match grain_type:
```

Comment:
Is this commented out because mypy doesn't support match yet? I wonder if a branch might be a better place for that to live? Either way, would you mind putting a Pivotal job in to merge the branch/uncomment it once the missing support is in?

Reply:
I may just delete it. It's not significantly better than using the if-ladder, really.